plasma/morevp_2.0.md

```diff
@@ -0,0 +1,109 @@
+3. Exchange is not finished with the `super_tx` at hand, so it chains the next tx to that: `E.chain(tx1, 1, tx2)`, indicating that 1st output `o5` of `tx1` has been consumed by a follow-up `tx2`.
+(Think of this as a "mixture" of a `startIFE` and `piggyback_output` message.)
+Observe that `tx2` consumes `o3` and `o5` as inputs: if this were a `startIFE` we'd require `E` to deliver inclusion proof of txs creating those `o3`, `o5`, but it is **only necessary** for `o3` - `o5` is chained instead of proven
+4. Here Bob and Charlie do what Alice did in (2/)
+
+If there were more txs to be chained `tx3`, `tx4`, ... the process would iterate naturally.
+
+`super_tx` here is `[tx1, tx2]`, it is a state transition from `o1`, `o2`, `o3` to `o4`, `o6`, `o7`.
+
+### Canonicity game and the `super_tx`
+
+#### General flow
+
+The "happened or not" of the `super_tx` is dependent on the canonicity of the individual chained txs.
+If both `tx1` and `tx2` are canonical, the `super_tx` happened in its entirety (`o4`, `o6`, `o7` should exit)
+If `tx2` turns out to be non-canonical, the `super_tx` happened only as far as `tx1` goes (`o3`, `o4`, `o5` should exit)
+If `tx1` turns out to be non-canonical, the `super_tx` didn't happen at all (`o1-3` should exit).
+
+#### Canonicity index
+
+One might think that this would require the contract to iterate over the chain of exits to change the status of the `super_tx`, as individual canonicities of the chained txs change.
+This is not the case - the exiting data can be represented in a way that allows this to be known without iterating.
+This is done by introducing the concept of `canonicity_index`.
+The "happened or not" of the `super_tx` is not represented by a boolean (canonical/non-canonical), but by an integer `n` meaning, `n` chained txs are "canonical".
+If there's an `n+1`th transaction then it is the first non-canonical in the chain.
+
+Using `canonicity_index`, one can organize the exiting funds of the `super_tx` so that on `processExits`, the cIFE "knows" which inputs and outputs to allow out.
+
+#### Requirements on competitors
+
+To ensure that the canonicity game makes sense, we are after the following property.
+There must be only 2 ways a chain of txs may have been attempted to be executed:
+ - by publishing by `E` in a cIFE
+ - by including in child chain blocks
+
+If `E` has double-signed some settlements, the version included in valid child chain blocks prevails and is published.
+If `E` didn't double-sign, the single chain of txs is the one in the cIFEs, will be canonical.
+
+In both cases the users have access to the funds.
+
+##### Proof for competitors required
+
+An important modification of the canonicity game we introduce for the chained txs is that submitting a competitor **requires an inclusion proof**.
+
+##### Unique `output_id` cIFE-ing
+
+In order to root out the possibility of two txs from two cIFEs being competitors, we need to keep a contract-global register of cIFE-ing `output_ids`.
```

Comment:
> In order to root out the possibility of two txs from two cIFEs being competitors

Why would this be a problem? Would they not just challenge non canonical and stop the chain at certain valid `n`?

Reply:
similar to the other reasons given in the comment just above - it simplifies stuff a lot, giving up some generality of the protocol.

But here there's also a second facet - allowing two such chains to be considered would require (honest) traders to piggyback on their outputs in various cIFEs, which would still be under exchange's restricted custody (think `o5` owned by `(E. B)`). In order to straighten this out we put this restriction in. Now, this `o5` can be freely piggybacked by `B` at its sole discretion.

Ah wait, there's a simpler argument: if `E` started a second cIFE from `tx2`, that would be difficult to resolve

But I agree this is the delicate portion of the whole construct and deserves more attention to explain in the doc

Reply:
ah okay, I re-read again so basically the goal here is to flag all `outputId` that is on the chain of cIFE. When we are doing this, we block possibility of other IFE/SE by checking `outputId` (?) so we only have one exit at one time? Then we don't need non-canonical game at all, is this correct?

Reply:
I think it's enough to force the check on cIFEs.

cIFEs are the "only IFEs" for settlement tx type (and possibly some withdrawal tx types must be treated as degenerate settlements here). Conversely, cIFEs are only there for settlement tx type.

Only within the realm of those cIFEs you restrict the `outputId`s.

>  we only have one exit at one time? 

better put, all the chains built in all cIFEs are disjoint. One way to see this is that in the event of a block withholding attack, there will be one cIFE per `{exchange, pair}`, chaining all the in-flight settlements for the pair.

> Then we don't need non-canonical game at all, is this correct?

hm, it is a degenerate canonicity game yes, but you still need to be able to challenge the settlements `E` claims in the cIFE with those that got included in the child chain - and the latter take priority. Hence, when someone has inclusion proofs for the competitors, the non-canonical txs in the cIFE become effectively useless.

Note that this heavily relies on the RC assumption. I wish there was a (still simple) exit game formulation, that could make us rely on that less, as a future improvement